### PR TITLE
Selectable diagnostic popover text

### LINKED ISSUE #12695: Can't click on diagnostic popovers
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

Normally, you should be able to mouse over a diagnostic popover, and click to go to diagnostic.

https://github.com/zed-industries/zed/assets/50590465/e0330fb7-fd8a-4e90-8426-3d290bda9732

However, in some languages, you can't. 

https://github.com/zed-industries/zed/assets/50590465/d2dd22c7-6441-4289-9ad3-fc1447cf1b26

One common thread I notice is you can click the diagnostic popover when there is an additional information popover present, but I'll look into this behavior a bit more. 

### Environment

Zed: v0.140.0 (Zed Dev 071270fe8845f8051e077b87958bec489dfd3be6)
OS: macOS 14.5.0
Memory: 16 GiB
Architecture: aarch64

### If applicable, add mockups / screenshots to help explain present your vision of the feature

_No response_

### If applicable, attach your `~/Library/Logs/Zed/Zed.log` file to this issue.

_No response_

### PR DESCRIPTION
**Pull request by effdotsh**
Release Notes:

- Added the ability to select and copy text from diagnostic popovers

- Fixed #12695

https://github.com/user-attachments/assets/b896bacf-ecbc-48f5-8228-a3821f0b1a4e

### CODE REVIEW

File: crates/editor/Cargo.toml [hunk 1]
```diff
@@ -28,6 +28,7 @@ test-support = [
 ]
 
 [dependencies]
+chrono = "0.4"
```
Comment (maxdeviant) [inline crates/editor/Cargo.toml:31]: We already have `chrono` in the workspace dependencies, so we should use `chrono.workspace = true` here.

Could we also keep this list sorted alphabetically?

File: crates/editor/src/hover_popover.rs [hunk 2]
```diff
@@ -299,12 +314,79 @@ fn show_hover(
...
+                                border_color = Some(status_colors.ignored_border);
+                            }
+                        };
+                        let settings = ThemeSettings::get_global(cx);
+                        let mut base_text_style = cx.text_style();
+                        base_text_style.refine(&TextStyleRefinement {
+                            font_family: Some(settings.ui_font.family.clone()),
+                            font_size: Some(settings.ui_font_size.into()),
+                            color: Some(cx.theme().colors().editor_foreground),
+                            // background_color: Some(background_color.unwrap()),
```
Comment (maxdeviant) [inline crates/editor/src/hover_popover.rs:360]: If we don't need this commented-out line anymore we should remove.

File: crates/editor/src/hover_popover.rs [hunk 3]
```diff
@@ -583,84 +673,55 @@ impl InfoPopover {
...
-        };
+        if let Some(border_color) = &self.border_color {
+            markdown_div = markdown_div
+                .border_1()
+                .border_color(border_color.clone())
+                .rounded_lg();
+        }
 
-        div()
+        let d = div()
```
Comment (maxdeviant) [inline crates/editor/src/hover_popover.rs:702]: Can we use a more descriptive variable name than `d` here?

File: crates/editor/Cargo.toml (hunk 1 above)
Comment (effdotsh) [reply crates/editor/Cargo.toml]: Whoops meant to remove that. Was just printing the time while debugging haha

**Comment by ConradIrwin** [comment]
@effdotsh Thanks for this!

One thing I'm on the fence about is whether we should interpret these diagnostics as markdown or not. It definitely seems that some languages output "markdown-like" syntax for diagnostics, but in general I think they're pretty plain text.

I made a change locally to try and differentiate code from plain text to make the rust ones more readable, but I'm not sure it's enough:

Before this PR:
Screenshot 2024-07-18 at 12 39 30

After (with monospace inline code):
Screenshot 2024-07-18 at 12 39 36

I do like that links are clickable (though we should underline them), but not sure what to do about the `s which are needed to parse the error message in this case.

Looking into how Go generates its diagnostics, it does not include any backticks around code samples, but its code may include backticks. In this case the problem is that I have "\`asda\`".

Before:
Screenshot 2024-07-18 at 12 55 02

After (with another idea to make code more visible):
Screenshot 2024-07-18 at 12 56 15

In conclusion, I think we should *not* run the markdown parser over these; but instead only run the link detector: https://github.com/zed-industries/zed/blob/b02f22428b23e158744a86c0e657c1c0045b6c32/crates/markdown/src/parser.rs#L35-L69; so in this case our markdown document can only contain Markdown::Text and Markdown::Link.

Probably the right API for that is to add a new constructor to the `Markdown` component (something like `Markdown::new_text()` that arranges for a different implementation of `parse_markdown` to be called.

**Comment by effdotsh** [comment]
@ConradIrwin I added a new_text constructor, is this closer to what you would expect?

image

image

**Comment by ConradIrwin** [comment]
Nice, thank you!

In the `parse_text` flow let's *not* pass the content to cmark. We can generate the Text/Link events ourselves as is done in the existing parser.

**Comment by effdotsh** [comment]
@ConradIrwin updated to not use cmark. I made a version that removed the brackets around the links, but given we're not processing any of the other markdown-y features I think it's better to leave the brackets untouched.

image

**Comment by ConradIrwin** [comment]
Nice! I think the final change needed is to revert the `soft_break_delim` option, which we don't need now we're not trying to parse this as markdown.

**Comment by effdotsh** [comment]
Done!
